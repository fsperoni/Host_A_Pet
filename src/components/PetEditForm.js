import React, { useState, useContext } from "react";
import Alert from "./Alert";
import HostAPetApi from "./Api";
import { TYPE } from "./HostAPet";
import UserContext from "../hooks/useUserContext";

/** Edit Pet form. */

const PetEditForm = ({pet, pets, setPets, setShowEditForm}) => {
  const { currentUser } = useContext(UserContext);
  const [formData, setFormData] = useState({
    ownerId: pet.ownerId,
    name: pet.name,
    type: pet.type,
    photo: pet.photo,
  });
  const [formErrors, setFormErrors] = useState([]);
  

  /** Handle form submit */
  const handleSubmit = async (evt) => {
    evt.preventDefault();
    const id = parseInt(evt.target.children[0].value);
    try {
      const result = await HostAPetApi.updatePet(currentUser.username, id, formData)
      const newPets = pets.filter(pet => pet.id !== id);
      setPets([...newPets, result.pet])
      setFormErrors([]);
      setShowEditForm(false);
    } catch (err) {
      setFormErrors(err);
    }
  }

  /** Handle form field changes */
  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData(data => ({ ...data, [name]: value }));
    setFormErrors([]);
  }

  /** Handle form cancel button */
  const handleCancel = (evt) => {
    evt.preventDefault();
    setFormErrors([]);
    setShowEditForm(false);
  }

  const generateOptions = () => {
    return TYPE.filter(t => t !== formData.type).map(animal => (
      <option key={animal} value={animal}>{animal}</option>
    ))
  }
  const options = generateOptions();
  return (
    <div className="EditPetForm mt-3">
      <div className="container">
        <h2 className="mb-3">Update pet info</h2>
        <div className="card">
          <div className="card-body">
            <form onSubmit={handleSubmit}>
              <div className="form-group">
                <label>Name:</label>
                <input name="name" className="form-control" onChange={handleChange}
                  value={formData.name}
                />
              </div>
              <div className="form-group">
                <label>Type:</label>
                <select name="type" className="form-select" onChange={handleChange}>
                  <option value={formData.type}>{formData.type}</option>
                  {options}
                </select>
              </div>
              <div className="form-group">
                <label>Photo:</label>
                <input name="photo" type="url" className="form-control" onChange={handleChange}
                  value={formData.photo}
                />
              </div>
              {formErrors.length>0 && <Alert type="danger" messages={formErrors} />}
              <div className="container mt-4">
              <button 
                className="btn btn-primary btn-block"
                onClick={handleSubmit}
              > <data value={pet.id} />
                Save
              </button>
              <button
                className="btn btn-warning btn-block ms-4"
                onClick={handleCancel}
              >
                Cancel
              </button>
            </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default PetEditForm;