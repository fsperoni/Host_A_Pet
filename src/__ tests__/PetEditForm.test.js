import { render } from "@testing-library/react";
import PetEditForm from "../components/PetEditForm";
import UserContext from "../hooks/useUserContext";

const {
  currentAdminUser,
  pets, pet
} = require("./_testCommon");

// smoke test
it('renders without crashing', function () {
  render(
      <UserContext.Provider
        value={{ currentUser: currentAdminUser(), function() { } }}>
        <PetEditForm setPets={function () { }} pets={pets()} 
          pet={pet()} setShowEditForm={function () {}}/>
      </UserContext.Provider>
  );
});

//snapshot test
it("matches snapshot", function () {
  const { asFragment } = render(
      <UserContext.Provider
        value={{ currentUser: currentAdminUser(), function() { } }}>
        <PetEditForm setPets={function () { }} pets={pets()} 
          pet={pet()} />
      </UserContext.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should have 'Update pet info' in the component.", function () {
  const { getByText } = render(
      <UserContext.Provider
        value={{ currentUser: currentAdminUser(), function() { } }}>
        <PetEditForm setPets={function () { }} pets={pets()} 
          pet={pet()} />
      </UserContext.Provider>
  );
  expect(getByText("Update pet info")).toBeInTheDocument();
});
