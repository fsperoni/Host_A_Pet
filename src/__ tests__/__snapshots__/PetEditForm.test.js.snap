// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="EditPetForm mt-3"
  >
    <div
      class="container"
    >
      <h2
        class="mb-3"
      >
        Update pet info
      </h2>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="form-group"
            >
              <label>
                Name:
              </label>
              <input
                class="form-control"
                name="name"
                required=""
                value="Garfield1"
              />
            </div>
            <div
              class="form-group"
            >
              <label>
                Type:
              </label>
              <select
                class="form-select"
                name="type"
                required=""
              >
                <option
                  value="Dog"
                >
                  Dog
                </option>
                <option
                  value="Cat"
                >
                  Cat
                </option>
              </select>
            </div>
            <div
              class="form-group"
            >
              <label>
                Photo:
              </label>
              <input
                class="form-control"
                name="photo"
                type="url"
                value="/static/media/dog.4b90f018c1cda3590605.jpeg"
              />
            </div>
            <div
              class="container mt-4"
            >
              <button
                class="btn btn-primary btn-block"
              >
                 
                <data
                  value="23"
                />
                Save
              </button>
              <button
                class="btn btn-warning btn-block ms-4"
              >
                Cancel
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
